server-metrics/src/main/scala/org/http4s/server/metrics/Metrics.scala

```diff
@@ -81,30 +73,25 @@ object Metrics {
             else if (code < 400) resp3xx.update(elapsed, TimeUnit.NANOSECONDS)
             else if (code < 500) resp4xx.update(elapsed, TimeUnit.NANOSECONDS)
             else resp5xx.update(elapsed, TimeUnit.NANOSECONDS)
-
-            cause match {
-              case End | Kill =>
-              case Error(_) =>
-                abnormal_termination.update(elapsed, TimeUnit.NANOSECONDS)
-            }
-            Halt(cause)
           }
-
-          Right(r.copy(body = body))
-
-       case Xor.Left(e)       =>
-          generalMetrics(method, elapsed)
-          resp5xx.update(elapsed, TimeUnit.NANOSECONDS)
-          service_failure.update(elapsed, TimeUnit.NANOSECONDS)
-          Left(e)
+        }.onError { cause =>
+          abnormal_termination.update(elapsed, TimeUnit.NANOSECONDS)
+          Stream.fail(cause)
+        }
+
+        r
+      }.leftMap { e =>
+        generalMetrics(method, elapsed)
+        resp5xx.update(elapsed, TimeUnit.NANOSECONDS)
+        service_failure.update(elapsed, TimeUnit.NANOSECONDS)
+        e
       }
     }
 
     Service.lift { req: Request =>
       val now = System.nanoTime()
       active_requests.inc()
-      new Task(service(req).get.map(onFinish(req.method, now)))
+      service(req).attempt.map(onFinish(req.method, now)).flatMap(_.fold(Task.fail, Task.now))
```

Comment:
yeah, totally agree - it felt wrong :)  I'll take a look at making it cleaner while I've got it in my head.